File: WorldFeed/Server/src/WorldFeed.Common/Client/Validators/TrendsClientParametersValidator.cs

namespace WorldFeed.Common.Client.Validators
{
    using WorldFeed.Common.Public.Parameters.TrendsClient;

    public interface ITrendsClientParametersValidator
    {
        void Validate(IGetTrendsLocationCloseToParameters parameters);

        void Validate(IGetTrendsAtParameters parameters);

        void Validate(IGetTrendsLocationParameters parameters);
    }

    public class TrendsClientParametersValidator : ITrendsClientParametersValidator
    {
        private readonly ITrendsClientRequiredParametersValidator requiredParametersValidator;

        public TrendsClientParametersValidator(ITrendsClientRequiredParametersValidator requiredParametersValidator)
        {
            this.requiredParametersValidator = requiredParametersValidator;
        }

        public void Validate(IGetTrendsLocationCloseToParameters parameters)
        {
            this.requiredParametersValidator.Validate(parameters);
        }

        public void Validate(IGetTrendsAtParameters parameters)
        {
            this.requiredParametersValidator.Validate(parameters);
        }

        public void Validate(IGetTrendsLocationParameters parameters)
        {
            this.requiredParametersValidator.Validate(parameters);
        }
    }
}

File: Sharebook/Client/src/app/core/Core/Client/Validators/TrendsClientParametersValidator.ts
import {IGetTrendsLocationCloseToParameters} from "../../../Public/Parameters/TrendsClient/GetTrendsLocationCloseToParameters";
import {IGetTrendsAtParameters} from "../../../Public/Parameters/TrendsClient/GetTrendsAtParameters";
import {IGetTrendsLocationParameters} from "../../../Public/Parameters/TrendsClient/GetTrendsLocationParameters";
import {ITrendsClientRequiredParametersValidator} from "./TrendsClientRequiredParametersValidator";
import {TrendsParameters} from "./parameters-types";
import {InjectionToken} from "@angular/core";

export interface ITrendsClientParametersValidator {
  validate(parameters: IGetTrendsLocationCloseToParameters): void;

  validate(parameters: IGetTrendsAtParameters): void;

  validate(parameters: IGetTrendsLocationParameters): void;
}

export const ITrendsClientParametersValidatorToken = new InjectionToken<ITrendsClientParametersValidator>('ITrendsClientParametersValidator', {
  providedIn: 'root',
  factory: () => new TrendsClientParametersValidator(),
});

export class TrendsClientParametersValidator implements ITrendsClientParametersValidator {
  private readonly _requiredParametersValidator: ITrendsClientRequiredParametersValidator;

  constructor(requiredParametersValidator: ITrendsClientRequiredParametersValidator) {
    this._requiredParametersValidator = requiredParametersValidator;
  }

  public validate(parameters: TrendsParameters): void {
    this._requiredParametersValidator.validate(parameters);
  }
}
